### PR TITLE
Rename Render* to Gpu*

### PR DESCRIPTION
**Pull request by kurtkuehnert**
# Objective

- alternative to #2895 
- as mentioned #2535, RenderDevice and RenderQueue should be renamed to GpuDevice and GpuQueue for better clarity and constituency 

## Solution

- This PR only focuses on renaming the type and variable names.
- renamed type names Render{Device, Queue, Adapter, AdapterInfo, Instance, Context} to GPU{Device, Queue, Adapter, AdapterInfo, Instance, Context}
- renamed variable names render_{device, queue, adapter, adapter_info, instance, context} to gpu_{device, queue, adapter, adapter_info, instance, context}

- maybe slightly controversial: I have additionally renamed a few occurrences of the variable name `device` and `queue`, which refer to a GPUDevice or GPUQueue, to `gpu_device` and `gpu_queue` for consistency's sake.

- additionally, I have type aliased the `CommandEncoder` to `GpuCommandEncoder`
---

### Changed

- Renamed the types Render{Device, Queue, Adapter, AdapterInfo, Instance, Context} to GPU{Device, Queue, Adapter, AdapterInfo, Instance, Context}

- Renamed `CommandEncoder` to `GpuCommandEncoder`

- The fields of the `RenderContext` have been renamed like so:

```rust
// old
pub struct RenderContext {
    pub device: RenderDevice,
    pub command_encoder: CommandEncoder,
}

// new
pub struct GpuContext {
    pub gpu_device: GpuDevice,
    pub gpu_command_encoder: GpuCommandEncoder,
}
```

## Migration Guide

Rename all occurrences of the types Render{Device, Queue, Adapter, AdapterInfo, Instance, Context} to GPU{Device, Queue, Adapter, AdapterInfo, Instance, Context}.

The fields of the `RenderContext` have been renamed like so:

```rust
// old
pub struct RenderContext {
    pub device: RenderDevice,
    pub command_encoder: CommandEncoder,
}

// new
pub struct GpuContext {
    pub gpu_device: GpuDevice,
    pub gpu_command_encoder: GpuCommandEncoder,
}
```

### CODE REVIEW

**Review by mockersf — APPROVED** [review]

**Comment by james7132** [comment]
Going to raise an objection to this rename. It's not guaranteed that the hardware doing the rendering is an actual GPU, It could easily just be a software renderer or a null device (in the case of headless environments). As mentioned above, it also adds a lot of churn to almost every rendering API, including many user-facing ones as well. Given these two reasons, I don't think the rename is well justified, as all it does is add a lot of busywork for contributors and users without gaining us that much clarity.

**Comment by devil-ira** [comment]
I'm seconding James on the points of churn and clarity.

It seems that wgpu also avoided the gpu prefix for the same reason despite the naming scheme in the WebGPU spec (e.g. [Device](https://docs.rs/wgpu/latest/wgpu/struct.Device.html) and [Adapter](https://docs.rs/wgpu/latest/wgpu/struct.Adapter.html)).

**Comment by kurtkuehnert** [comment]
I do understand your point. 
Although, I personally do not think that there is any problem with classifying a software rendering backend to wgpu as a GPUAdapter, etc.

As I see it we have got three options: 
1. stick with the current `Render` prefix 
- no change
- these types are not used for rendering alone, but also for GPUCompute
- until now we do not offer any high-level compute abstraction but in the future, these names will be very confusing
2. go through with this PR and adopt the `GPU` prefix 
- all users of these types will have to search and replace a couple of types and maybe some variable names
- we run into the issue with the GPU - null device - software renderer 
3. we could also just use the same naming scheme as wgpu does and rename the wgpu types (e.g. WgpuAdapter) 
- this could cause some confusion, but since we are already abstracting a way most of the raw wgpu interface this could also be good solution
- however, this would also result in a rename of types and variables, that will bubble up to our users

I would prefer solution 2 or 3. with 3 being slightly more elegant in my opinion. What do you think?

**Comment by kurtkuehnert** [comment]
Also the further we delay this decision the more user code we break. I should have probably redone this PR last year then we would not have as many users depending on these features.

**Comment by cart** [comment]
I still consider churn to be acceptable for most Bevy changes at this point. If we agree something is "the right call" then we should do it, even if it breaks things (especially for things like renames which are straightforward migrations).

So I think the "clarity" point is what we should discuss.

Some arguments in favor of the change:
1. WGPU is GPU api. Anyone building a "real" CPU renderer wouldn't go through a wgpu-like interface. The design was entirely informed by the constraints of gpus and they use terminology rooted in GPU history (ex: shader, bind group, etc).
2. RenderDevice is "too specific". Not all gpu-compute operations will relate to rendering (ex: users might do simulations or other app logic on the gpu).

My prediction for the reason WGPU avoided the GPU prefix is that it is redundant / implied within the context of WGPU. Within the context of Bevy, GPU cannot be implied context for a Device, as we deal with many devices in many contexts (input devices, display devices, etc). Same goes for Adapter (ex: network adapter).

If we do roll with "Gpu" terminology, we might even want to forgo the "Device" term and just call it `Gpu`. Thats slightly controversial because it strips out a wgpu term entirely. But `GpuDevice` feels a bit redundant to me, as GPU already implies device.

**Comment by kurtkuehnert** [comment]
I totally agree with 1 and 2.
Now that I have thought more about it, we should not adopt the wgpu naming convention, due to the reasons that you have mentioned.

But I do not agree that we rename the `Device` to `Gpu`.
WebGPU calls the wgpu `Instance` `GPU`. If we were to rename the `Device` to `Gpu` that would cause a lot of confusion.
Here is a nice comparison I have found between the terms used by all the different graphics APIs https://alain.xyz/blog/comparison-of-modern-graphics-apis.

I am still in favor of this rename, but I will change the prefix to `Gpu` instead of `GPU` to match the rust convention.

Maybe we should do a poll in the bevy render-dev channel?

File: crates/bevy_render/src/render_resource/storage_buffer.rs [hunk 1]
```diff
@@ -101,26 +101,26 @@ impl<T: ShaderType + WriteInto> StorageBuffer<T> {
         self.label.as_deref()
     }
 
-    /// Queues writing of data from system RAM to VRAM using the [`RenderDevice`](crate::renderer::RenderDevice)
-    /// and the provided [`RenderQueue`](crate::renderer::RenderQueue).
+    /// Queues writing of data from system RAM to VRAM using the [`GpuDevice`](crate::renderer::GpuDevice)
+    /// and the provided [`GPUQueue`](crate::renderer::GpuQueue).
```
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/storage_buffer.rs:105]:
```suggestion
    /// and the provided [`GpuQueue`](crate::renderer::GpuQueue).
```

File: crates/bevy_render/src/render_resource/uniform_buffer.rs [hunk 2]
```diff
@@ -99,23 +99,23 @@ impl<T: ShaderType + WriteInto> UniformBuffer<T> {
         self.label.as_deref()
     }
 
-    /// Queues writing of data from system RAM to VRAM using the [`RenderDevice`](crate::renderer::RenderDevice)
-    /// and the provided [`RenderQueue`](crate::renderer::RenderQueue), if a GPU-side backing buffer already exists.
+    /// Queues writing of data from system RAM to VRAM using the [`GPUDevice`](crate::renderer::GpuDevice)
+    /// and the provided [`GPUQueue`](crate::renderer::GpuQueue), if a GPU-side backing buffer already exists.
```
Comment (IceSentry) [inline crates/bevy_render/src/render_resource/uniform_buffer.rs:103]:
```suggestion
    /// Queues writing of data from system RAM to VRAM using the [`GpuDevice`](crate::renderer::GpuDevice)
    /// and the provided [`GpuQueue`](crate::renderer::GpuQueue), if a GPU-side backing buffer already exists.
```

Comment (IceSentry) [reply crates/bevy_render/src/render_resource/uniform_buffer.rs:103]: There's a couple of those in the PR, so I won't add a suggestion everywhere, but for consistency it's probably better to use the same capitalization as the struct.

**Review by IceSentry — APPROVED** [review]
Overall I like it, especially based on cart's comment and the discord poll. There's a few instances of GPU vs Gpu that should be fixed, but other than that and assuming the merge conflicts are fixed LGTM

**Comment by kurtkuehnert** [comment]
It seems that most people are in favor of this change.
Ideally, this PR will be followed by the #7000 (or multiple smaller PRs) bevy_gpu/bevy_render split.

I still would like to hear your opinions on some decisions related to the entire refactor. 

1. The Render/GpuContext is only used during the RenderGraph. I think it should be moved there.
Here I am wondering whether the Render prefix is the better name.

2. Currently we use multiple different Prefixes for GPU-related stuff: Render, Gpu and Wgpu. 
This PR changes Render to Gpu. I believe that we should also rename the Wgpu* types (like WgpuSettings, WgpuLimits, etc.) to Gpu*. I can do this here or as part of a separate PR.

3. With the type renames complete, the next step would be to rename the `render_resource` module to `gpu_resource`.

4. The refactor of #7000 went surprisingly smoothly, except for three minor details.
4.1 The PipelineCache::extract_shaders method depends on the Extract Param of the render module. Thus I had to move it out of bevy_gpu. I am still unsure where to put this code though.
4.2 The AsBindGroup trait is a render abstraction and thus had to be moved out of bevy_gpu. Same as above where should I put this?
4.3 The SpecializedMeshPipeline was moved out of the bevy_gpu::gpu_resource module into the bevy_render::mesh module. This is more in line with what we do for other specialized pipelines as well.

5. To make reviewing easier I could try to separate the refactor (file renames, type moves, etc.) into multiple smaller PRs before I transfer everything into the `bevy_gpu` crate. This will create a lot more changes though (especially the `use` statements will change multiple times). Alternatively, the split can be performed as one large PR (eg. #7000). 

6. Currently, there is still an open TODO in the `render_resource` module stating: 
> // TODO: decide where re-exports should go

We can keep them where they are (`gpu_resource`), or elevate them into the root `bevy_gpu` module. 
Whatever we choose we should probably do the same for the GPU resources to make importing easier.

```rust
use bevy::gpu::Buffer;
vs
use bevy::gpu::gpu_resource::Buffer;
```

7. I think we should use the opportunity to hide all wgpu specifics inside the `bevy_gpu` crate, and only reexport the stuff we have to. `bevy_render` should no longer depend on both wgpu and encase.

**Pull request closed without merging**